Mathlib/Combinatorics/Graph/Basic.lean

```diff
@@ -0,0 +1,325 @@
+  by_contra! hcon
+  obtain ⟨x', hx'⟩ := hx
+  obtain rfl := hy.eq_of_isLink_of_ne_left hx' hcon.1.symm
+  obtain rfl := hz.eq_of_isLink_of_ne_left hx' hcon.2.1.symm
+  exact hcon.2.2 rfl
+
+/-- `G.IsLoopAt e x` means that both ends of `e` are equal to `x`. -/
+def IsLoopAt (G : Graph α β) (e : β) (x : α) : Prop := G.IsLink e x x
+
+@[simp]
+lemma isLink_self_iff : G.IsLink e x x ↔ G.IsLoopAt e x := Iff.rfl
+
+lemma IsLoopAt.inc (h : G.IsLoopAt e x) : G.Inc e x :=
+  IsLink.inc_left h
+
+lemma IsLoopAt.eq_of_inc (h : G.IsLoopAt e x) (h' : G.Inc e y) : x = y := by
+  obtain rfl | rfl := h'.eq_or_eq_of_isLink h <;> rfl
+
+@[simp]
+lemma IsLoopAt.edge_mem (h : G.IsLoopAt e x) : e ∈ E(G) :=
+  h.inc.edge_mem
+
+@[simp]
+lemma IsLoopAt.vertex_mem (h : G.IsLoopAt e x) : x ∈ V(G) :=
+  h.inc.vertex_mem
+
+/-- `G.IsNonloopAt e x` means that `e` is an edge from `x` to some `y ≠ x`,
+or equivalently that `e` is incident with `x` but not a loop at `x` -
+see `Graph.isNonloopAt_iff_inc_not_isLoopAt`. -/
+def IsNonloopAt (G : Graph α β) (e : β) (x : α) : Prop := ∃ y ≠ x, G.IsLink e x y
+
+lemma IsNonloopAt.inc (h : G.IsNonloopAt e x) : G.Inc e x :=
+  h.choose_spec.2.inc_left
+
+@[simp]
+lemma IsNonloopAt.edge_mem (h : G.IsNonloopAt e x) : e ∈ E(G) :=
+  h.inc.edge_mem
+
+@[simp]
+lemma IsNonloopAt.vertex_mem (h : G.IsNonloopAt e x) : x ∈ V(G) :=
+  h.inc.vertex_mem
+
+lemma IsLoopAt.not_isNonloop_at (h : G.IsLoopAt e x) (y : α) : ¬ G.IsNonloopAt e y := by
+  rintro ⟨z, hyz, hy⟩
+  rw [← h.eq_of_inc hy.inc_left, ← h.eq_of_inc hy.inc_right] at hyz
+  exact hyz rfl
+
+lemma IsNonloopAt.not_isLoopAt (h : G.IsNonloopAt e x) (y : α) : ¬ G.IsLoopAt e y :=
+  fun h' ↦ h'.not_isNonloop_at x h
+
+lemma isNonloopAt_iff_inc_not_isLoopAt : G.IsNonloopAt e x ↔ G.Inc e x ∧ ¬ G.IsLoopAt e x :=
+  ⟨fun h ↦ ⟨h.inc, h.not_isLoopAt _⟩, fun ⟨⟨y, hy⟩, hn⟩ ↦ ⟨y, mt (fun h ↦ h ▸ hy) hn, hy⟩⟩
+
+lemma isLoopAt_iff_inc_not_isNonloopAt : G.IsLoopAt e x ↔ G.Inc e x ∧ ¬ G.IsNonloopAt e x := by
+  simp +contextual [isNonloopAt_iff_inc_not_isLoopAt, iff_def, IsLoopAt.inc]
+
+lemma Inc.isLoopAt_or_isNonloopAt (h : G.Inc e x) : G.IsLoopAt e x ∨ G.IsNonloopAt e x := by
+  simp [isNonloopAt_iff_inc_not_isLoopAt, h, em]
+
+/-! ### Adjacency -/
+
+/-- `G.Adj x y` means that `G` has an edge from `x` to `y`. -/
+def Adj (G : Graph α β) (x y : α) : Prop := ∃ e, G.IsLink e x y
+
+lemma Adj.symm (h : G.Adj x y) : G.Adj y x :=
+  ⟨_, h.choose_spec.symm⟩
+
+lemma adj_comm : G.Adj x y ↔ G.Adj y x :=
+  ⟨Adj.symm, Adj.symm⟩
+
+@[simp]
+lemma Adj.left_mem (h : G.Adj x y) : x ∈ V(G) :=
+  h.choose_spec.left_mem
+
+@[simp]
+lemma Adj.right_mem (h : G.Adj x y) : y ∈ V(G) :=
+  h.symm.left_mem
+
+lemma IsLink.adj (h : G.IsLink e x y) : G.Adj x y :=
+  ⟨e, h⟩
+
+/-! ### Extensionality -/
+
+/-- A constructor for `Graph` in which the edge set is inferred from the incidence predicate
+rather than being supplied explicitly. -/
+@[simps]
+protected def mk' (vertexSet : Set α) (IsLink : β → α → α → Prop)
```

Comment:
Do we actually benefit from this constructor, or could we make `edgeSet := {e | ∃ x y, IsLink e x y}` a default in the regular constructor?

Reply:
Is there a way to fill in the default proof of edge_mem_iff_exists_isLink as fun e => Iff.rfl if doing things this way?

Reply:
Yes, proof fields too can have default arguments

Reply:
Yeah, see for example the constructor for `lt` on preorders and a property describing it `lt_iff_le_not_le`: https://github.com/leanprover-community/mathlib4/blob/6534503fc2ed0aa28f544b96125ee6f97a26e30f/Mathlib/Order/Defs/PartialOrder.lean#L38-L43

Reply:
I couldn't get this to work. 

```
edgeSet : Set β := {e | ∃ x y, IsLink e x y}
  edge_mem_iff_exists_isLink : ∀ e, e ∈ edgeSet ↔ ∃ x y, IsLink e x y := fun e ↦ Iff.rfl
``` 

gives 
```
type mismatch
  Iff.rfl
has type
  ?m.234 ↔ ?m.234 : Prop
but is expected to have type
  e ∈ edgeSet ↔ ∃ x y, IsLink e x y : Prop
```

Reply:
It needs to be hidden inside a `by` block, as otherwise it is immediately evaluated

Reply:
Ok, the last push removes it.